### PR TITLE
add AutoML regression testing mechanism

### PR DESCRIPTION
**Pull request by jppgks**
This introduces 2 ways of verifying that a change in AutoML logic did not introduce errors in type inference on known datasets:
1) Using the pytest in test_type_inference.py
2) The expected types for each of the tested datasets are generated with AutoML by scripts/update_golden_types.py and checked in to the folder golden/. Running scripts/update_golden_types.py should result in a clean git diff.

### CODE REVIEW

File: tests/regression_tests/automl/test_type_inference.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import json
+
+import pytest
+
+from ludwig.automl.automl import create_auto_config
+from tests.regression_tests.automl.utils import get_dataset_golden_types_path, get_dataset_object, REGISTRY
+
+
+@pytest.mark.parametrize("dataset_name", REGISTRY)
+def test_ludwig_auto_config(dataset_name):
```
Comment (justinxzhao) [inline tests/regression_tests/automl/test_type_inference.py:10]: nit: Add "type inference" to the test name, i.e. `test_auto_type_inference_regression`

File: tests/regression_tests/automl/utils.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+from pathlib import Path
+
+from ludwig.datasets import dataset_registry
+from ludwig.datasets.base_dataset import BaseDataset
+
+REGISTRY = {"adult_census_income", "mnist"}
```
Comment (justinxzhao) [inline tests/regression_tests/automl/utils.py:6]: nit: rename to `TEST_DATASET_REGISTRY` and add a comment about what this registry is used for, i.e. "Subset of Ludwig Zoo datasets use for automl type inference regression tests."

File: tests/regression_tests/automl/utils.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def get_dataset_golden_types_path(dataset_name: str) -> str:
```
Comment (justinxzhao) [inline tests/regression_tests/automl/utils.py:9]: nit: Add docstring.

File: tests/regression_tests/automl/utils.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    return str(Path(__file__).resolve().parent / "golden" / f"{dataset_name}.types.json")
+
+
+def get_dataset_object(dataset_name: str) -> BaseDataset:
```
Comment (justinxzhao) [inline tests/regression_tests/automl/utils.py:13]: nit: Add docstring.

**Review by justinxzhao — COMMENTED** [review]
Looks like a solid regression test setup for type inference checking to me. Just a few nits.

File: tests/regression_tests/automl/utils.py (hunk 2 above)
Comment (w4nderlust) [reply tests/regression_tests/automl/utils.py:6]: agree, there are very many registries in the codebase so this helps dismbiguate ;)

**Comment by w4nderlust** [comment]
After nits are addressed, this looks good to me.
Will make sure to mark this as SLOW, or with a new specific decorator.
Downloading Mnist (and other datasets like amazon reviews, yahoo etc.) could be quite slow and potentially flaky depending on connections speed, so would not include these tests as default, but suggest to run them only when modifying automl type inference

**Review by justinxzhao — APPROVED** [review]
LGTM after some minor comments!

**Comment by jppgks** [comment]
@justinxzhao existing comments are addressed in the latest changes. Can you resolve if you're happy with the changes?

**Comment by justinxzhao** [comment]
@w4nderlust Agreed that minimizing flakiness is really important. At the moment, we have a few non-`@slow` tests in `test_text_encoders` that downloads weights from huggingface (O(10-100 MB)). We run into some issues with the tests rarely, but they have been largely stable. Perhaps we should rethink some of those tests as well.

@jppgks As Piero suggests, let's introduce a new decorator for these automl type inference regression tests, and only run them if there are changes to any files under `ludwig/automl`.

https://github.community/t/is-it-possible-to-run-the-job-only-when-a-specific-file-changes/115484

**Comment by jppgks** [comment]
Thanks for clarifying, Justin! On it 🙂

**Review by justinxzhao — APPROVED** [review]

**Comment by jppgks** [comment]
As agreed with @justinxzhao, will work in separate PR on the specific decorator/conditionally running test.

**Pull request merged**